packages/plugins/typescript/apollo-client-helpers/src/index.ts

```diff
@@ -0,0 +1,84 @@
+import { Types, PluginValidateFn, PluginFunction } from '@graphql-codegen/plugin-helpers';
+import { GraphQLSchema, isInterfaceType, isObjectType } from 'graphql';
+import { extname } from 'path';
+import { ApolloClientHelpersConfig } from './config';
+
+export const plugin: PluginFunction<ApolloClientHelpersConfig> = (
+  schema: GraphQLSchema,
+  documents: Types.DocumentFile[],
+  config: ApolloClientHelpersConfig
+) => {
+  const results: Types.ComplexPluginOutput[] = [];
+
+  results.push(generateTypePoliciesSignature(schema, config));
+
+  return {
+    prepend: results.reduce((prev, r) => [...prev, ...(r.prepend || [])], []),
+    append: results.reduce((prev, r) => [...prev, ...(r.append || [])], []),
+    content: results.map(r => r.content).join('\n'),
+  };
+};
+
+function generateTypePoliciesSignature(
+  schema: GraphQLSchema,
+  config: ApolloClientHelpersConfig
+): Types.ComplexPluginOutput {
+  const typeMap = schema.getTypeMap();
+  const perTypePolicies: string[] = [];
+  const typedTypePolicies = Object.keys(typeMap).reduce((prev, typeName) => {
+    const type = typeMap[typeName];
+
+    if (!typeName.startsWith('__') && (isObjectType(type) || isInterfaceType(type))) {
+      const fieldsNames = Object.keys(type.getFields()).filter(f => !f.startsWith('__'));
+      const keySpecifierVarName = `${typeName}KeySpecifier`;
+      const fieldPolicyVarName = `${typeName}FieldPolicy`;
+
+      perTypePolicies.push(
+        `export type ${keySpecifierVarName} = (${fieldsNames
+          .map(f => `'${f}'`)
+          .join(' | ')} | ${keySpecifierVarName})[];`
+      );
+
+      perTypePolicies.push(`export type ${fieldPolicyVarName} = {
+${fieldsNames.map(fieldName => `\t${fieldName}?: FieldPolicy<any> | FieldReadFunction<any>`).join(',\n')}
+};`);
+
+      return {
+        ...prev,
+        [typeName]: `{
+\t\tkeyFields?: false | ${keySpecifierVarName} | (() => undefined | ${keySpecifierVarName}),
+\t\tqueryType?: true,
+\t\tmutationType?: true,
+\t\tsubscriptionType?: true,
+\t\tfields?: ${fieldPolicyVarName},
+\t}`,
+      };
+    }
+
+    return prev;
+  }, {} as Record<string, string>);
+
+  const rootContent = `export type TypedTypePolicies = TypePolicies & {${Object.keys(typedTypePolicies)
+    .map(typeName => `\n\t${typeName}?: ${typedTypePolicies[typeName]}`)
```

Comment:
would it be possible to get an option for removing the `?` here, making it so we _have_ to specify all types?

Reply:
except for fields defined in `schema`:

```graphql
schema {
  query: Query
  mutation: Mutation
}
```

https://graphql.org/learn/schema/#the-query-and-mutation-types

Reply:
Working on those fixes here: https://github.com/dotansimha/graphql-code-generator/pull/4979

Reply:
> Yay, this is awesome! Thoughts on the directives and stuff I mentioned in #4700 and possible failing if I query for a type but do not include all of its keys? A directive could be used to generate the config (or at least their `keyFields`) instead of just verifying the type 🙂

@SimenB I'm trying to avoid usage of directives / changes to the schema to manipulate codegen behaviour...